File: backend/src/main/java/com/tomy/artifyme/artwork/ArtworkService.java
package com.tomy.artifyme.artwork;

import lombok.RequiredArgsConstructor;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;


import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import org.springframework.data.domain.Pageable;

@Service
@RequiredArgsConstructor
public class ArtworkService {

    private final ArtworkRepository artworkRepository;

    @SuppressWarnings("null")
    public ResponseEntity<ArtworkResponse> saveArtwork(ArtworkRequest request) {
        try {
            // Debug print to show the received request
            System.out.println("Request received: " + request);

            // Validate request parameters
            if (!StringUtils.hasText(request.getUserEmail()) ||
                    !StringUtils.hasText(request.getSketchedImage()) ||
                    !StringUtils.hasText(request.getAiImage()) || 
                    !StringUtils.hasText(request.getTitle()) || 
                    !StringUtils.hasText(request.getDescription()) ||
                    request.getPaths() == null || request.getPaths().isEmpty())
                    {
                throw new IllegalArgumentException("User email, sketched image, AI image, description , title and SVG paths are required!");
            }

            // Debug print to show validation success
            System.out.println("Request parameters validated successfully");

            // Create Artwork entity from the request and set creation date
            Artwork artwork = Artwork.builder()
                    .userEmail(request.getUserEmail())
                    .sketchedImage(request.getSketchedImage())
                    .aiImage(request.getAiImage())
                    .creationDateTime(LocalDateTime.now()) // Set creation date
                    .description(request.getDescription())
                    .paths(request.getPaths())
                    .title(request.getTitle())
                    .build();

            // Debug print to show the created artwork
            System.out.println("Artwork created: " + artwork);

            // Save the artwork
            artworkRepository.save(artwork);

            // Debug print to indicate successful artwork save
            System.out.println("Artwork saved successfully");
            // Return success message
          
            return ResponseEntity.ok(ArtworkResponse.builder()
                    .message("Artwork saved successfully")
                    .id(artwork.getId())
                    .build());
        } catch (Exception e) {
            // Print stack trace for debugging
            e.printStackTrace();

           return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(ArtworkResponse.builder()
                .message("An unexpected error occurred")
                .id(null)
                .build());
        }
    }

    public ResponseEntity<PageResponse> getArtworksWithPagination(
     Integer pageNumber,
    Integer pageSize,
    String userEmail) {

        try{
            // Create a pageable object for pagination
            Pageable pageable = PageRequest.of(pageNumber - 1, pageSize);
        
            // Perform paginated query with user email filter if provided
            Page<Artwork> page;
            if (userEmail != null && !userEmail.isEmpty()) {
                page = artworkRepository.findByUserEmailOrderByCreationDateTimeDesc(userEmail, pageable);
            } else {
                page = artworkRepository.findAll(pageable);
            }
        
            // Determine if there are more pages available
            boolean isNext = page.hasNext();
        
            // Create a PageResponse object
            return ResponseEntity.ok(PageResponse.builder()
                    .message("Success getting results")
                    .isNext(isNext)
                    .content(page.getContent())
                    .build());

        }catch(Exception e)
        {
              // Print stack trace for debugging
              e.printStackTrace();

              return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(PageResponse.builder()
                   .message("An unexpected error occurred: " + e)
                   .isNext(false)
                   .content(new ArrayList<>())
                   .build());
        }
    }
    
    public ResponseEntity<Object>getArtworkById(String id) {
        try {
            if (id == null || id.isEmpty()) {
                throw new IllegalArgumentException("ID is required!");
            }

            // Fetch artwork from the repository by ID
            Optional<Artwork> artworkOptional = artworkRepository.findById(id);

            if (artworkOptional.isPresent()) {
                Artwork artwork = artworkOptional.get();
                Map<String, Object> successResponse = new HashMap<>();
                successResponse.put("artwork", artwork);
                successResponse.put("message", "Artwork found");
                // If artwork is found, return it
                return ResponseEntity.ok(successResponse);
            } else {
                // If artwork is not found, return a not found response
                Map<String, Object> errorResponse = new HashMap<>();
                errorResponse.put("message", "Artwork not found");
                return ResponseEntity.status(404).body(errorResponse);
            }
        } catch (Exception e) {
            // Log exception for debugging
            e.printStackTrace();

            // If any other exception occurs, return an error response
            Map<String, Object> errorResponse = new HashMap<>();
            errorResponse.put("message", "Internal Server Error: " + e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
        }
    }

    public ResponseEntity<Object> deleteArtwork(String id) {
        try {
            // Check if ID is provided
            if (id == null || id.isEmpty()) {
                throw new IllegalArgumentException("ID is required!");
            }

            // Find the artwork by ID
            Optional<Artwork> artworkOptional = artworkRepository.findById(id);

            if (artworkOptional.isPresent()) {
                // If artwork is found, delete it
                artworkRepository.deleteById(id);

                // Return success response
                Map<String, Object> successResponse = new HashMap<>();
                successResponse.put("message", "Artwork deleted successfully");
                return ResponseEntity.ok(successResponse);
            } else {
                // If artwork is not found, return not found response
                Map<String, Object> errorResponse = new HashMap<>();
                errorResponse.put("message", "Artwork not found");
                return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorResponse);
            }
        } catch (Exception e) {
            // Log exception for debugging
            e.printStackTrace();

            // Return error response for any exception
            Map<String, Object> errorResponse = new HashMap<>();
            errorResponse.put("message", "Internal Server Error: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
        }
    }

    @SuppressWarnings("null")
    public ResponseEntity<Object> updateArtwork(String id, ArtworkRequest request) {
        try {
            // Find the artwork by its ID
            Optional<Artwork> optionalArtwork = artworkRepository.findById(id);
            if (!optionalArtwork.isPresent()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(ArtworkResponse.builder()
                                .message("Artwork not found")
                                .build());
            }
    
            Artwork artwork = optionalArtwork.get();
    
            // Update the artwork fields based on the request
            if (StringUtils.hasText(request.getTitle())) {
                artwork.setTitle(request.getTitle());
            }
            if (StringUtils.hasText(request.getDescription())) {
                artwork.setDescription(request.getDescription());
            }
            if (request.getSketchedImage() != null) {
                artwork.setSketchedImage(request.getSketchedImage());
            }
            if (request.getAiImage() != null) {
                artwork.setAiImage(request.getAiImage());
            }
            if (request.getPaths() != null && !request.getPaths().isEmpty()) {
                artwork.setPaths(request.getPaths());
            }
    
            // Save the updated artwork
            artworkRepository.save(artwork);
    
            // Debug print to indicate successful artwork update
            System.out.println("Artwork updated successfully: " + artwork);
    
            // Return success message
            return ResponseEntity.ok(ArtworkResponse.builder()
                    .message("Artwork updated successfully")
                    .id(artwork.getId())
                    .build());
        } catch (Exception e) {
            // Print stack trace for debugging
            e.printStackTrace();
    
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ArtworkResponse.builder()
                            .message("An unexpected error occurred")
                            .id(null)
                            .build());
        }
    }
    
    }

File: backend/Services/ArtworkService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using ArtifyMe.Models;
using ArtifyMe.Repositories.Interfaces;
using ArtifyMe.Models.DTOs;
using ArtifyMe.Services.Interfaces;


namespace ArtifyMe.Services
{
    public class ArtworkService : IArtworkService
    {
        private readonly IArtworkRepository _artworkRepository;

        public ArtworkService(IArtworkRepository artworkRepository)
        {
            _artworkRepository = artworkRepository;
        }

        //Save Artwork
        public async Task<IActionResult> SaveArtwork(ArtworkRequest request)
        {
            try
            {
                // Validate request parameters
                if (string.IsNullOrWhiteSpace(request.UserEmail) ||
                    string.IsNullOrWhiteSpace(request.SketchedImage) ||
                    string.IsNullOrWhiteSpace(request.AiImage) ||
                    string.IsNullOrWhiteSpace(request.Title) ||
                    string.IsNullOrWhiteSpace(request.Description) ||
                    request.Paths == null || !request.Paths.Any())
                {
                    return new BadRequestObjectResult(new ArtworkResponse
                    {
                        Message = "User email, sketched image, AI image, description, title, and paths are required!",
                        Id = null
                    });
                }

                // Create Artwork entity from the request
                var artwork = new Artwork
                {
                    UserEmail = request.UserEmail,
                    SketchedImage = request.SketchedImage,
                    AiImage = request.AiImage,
                    CreationDateTime = DateTime.UtcNow,
                    Description = request.Description,
                    Paths = request.Paths.Select(p => new PathData
                    {
                        Color = p.Color,
                        Path = p.Path,
                        Size = p.Size
                    }).ToList(),
                    Title = request.Title
                };

                // Save the artwork to the repository
                await _artworkRepository.InsertArtworkAsync(artwork);

                // Return success response
                return new OkObjectResult(new ArtworkResponse
                {
                    Message = "Artwork saved successfully",
                    Id = artwork.Id
                });
            }
            catch (Exception ex)
            {
                // Handle exceptions and return internal server error
                Console.WriteLine(ex);
                return new ObjectResult(new ArtworkResponse
                {
                    Message = "An unexpected error occurred",
                    Id = null
                })
                {
                    StatusCode = StatusCodes.Status500InternalServerError
                };
            }
        }

        // Get artworks with pagination
public async Task<IActionResult> GetArtworksWithPagination(int pageNumber, int pageSize, string userEmail)
{
    try
    {
        // Fetch one extra artwork to check if there's a next page
        var artworks = await _artworkRepository.GetArtworksWithPagination(pageNumber, pageSize + 1, userEmail);

        // Determine if there are more pages based on the extra item
        bool isNext = artworks.Count > pageSize;
        var artworksToReturn = artworks.Take(pageSize).ToList();

        // Create and return page response
        return new OkObjectResult(new PageResponse
        {
            Message = "Success getting results",
            IsNext = isNext,
            Content = artworksToReturn
        });
    }
    catch (Exception ex)
    {
        // Handle exceptions and return internal server error
        Console.WriteLine(ex);
        return new ObjectResult(new PageResponse
        {
            Message = $"An unexpected error occurred: {ex.Message}",
            IsNext = false,
            Content = new List<Artwork>()
        })
        {
            StatusCode = StatusCodes.Status500InternalServerError
        };
    }
}
   
        // Get artwork by ID
        public async Task<IActionResult> GetArtworkById(string id)
        {
            try
            {
                // Validate ID
                if (string.IsNullOrEmpty(id))
                {
                    return new BadRequestObjectResult(new { Message = "ID is required!" });
                }

                // Fetch artwork from the repository
                var artwork = await _artworkRepository.FindByIdAsync(id);

                // Check if artwork exists
                if (artwork == null)
                {
                    return new NotFoundObjectResult(new { Message = "Artwork not found" });
                }

                // Return the artwork
                return new OkObjectResult(new { Artwork = artwork, Message = "Artwork found" });
            }
            catch (Exception ex)
            {
                // Handle exceptions and return internal server error
                Console.WriteLine(ex);
                return new ObjectResult(new { Message = $"Internal Server Error: {ex.Message}" })
                {
                    StatusCode = StatusCodes.Status500InternalServerError
                };
            }
        }

        // Delete artwork by ID
        public async Task<IActionResult> DeleteArtwork(string id)
        {
            try
            {
                // Validate ID
                if (string.IsNullOrEmpty(id))
                {
                    return new BadRequestObjectResult(new { Message = "ID is required!" });
                }

                // Check if artwork exists
                var artwork = await _artworkRepository.FindByIdAsync(id);
                if (artwork == null)
                {
                    return new NotFoundObjectResult(new { Message = "Artwork not found" });
                }

                // Delete artwork
                await _artworkRepository.DeleteArtworkAsync(id);

                // Return success message
                return new OkObjectResult(new { Message = "Artwork deleted successfully" });
            }
            catch (Exception ex)
            {
                // Handle exceptions and return internal server error
                Console.WriteLine(ex);
                return new ObjectResult(new { Message = $"Internal Server Error: {ex.Message}" })
                {
                    StatusCode = StatusCodes.Status500InternalServerError
                };
            }
        }

        // Update artwork
       public async Task<IActionResult> UpdateArtwork(string id, ArtworkRequest request)
    {
    try
    {
        // Find the artwork by ID
        var artwork = await _artworkRepository.FindByIdAsync(id);
        if (artwork == null)
        {
            return new NotFoundObjectResult(new ArtworkResponse
            {
                Message = "Artwork not found"
            });
        }

        // Update artwork fields based on the request
        artwork.Title = request.Title ?? artwork.Title;
        artwork.Description = request.Description ?? artwork.Description;
        artwork.SketchedImage = request.SketchedImage ?? artwork.SketchedImage;
        artwork.AiImage = request.AiImage ?? artwork.AiImage;
        if (request.Paths != null && request.Paths.Any())
        {
            // Convert request paths to PathData
            artwork.Paths = request.Paths.Select(p => new PathData 
            { 
                Color = p.Color, 
                Path = p.Path, 
                Size = p.Size 
            }).ToList();
        }

        // Save the updated artwork
        await _artworkRepository.UpdateArtworkAsync(artwork);

        // Return success message
        return new OkObjectResult(new ArtworkResponse
        {
            Message = "Artwork updated successfully",
            Id = artwork.Id
        });
    }
    catch (Exception ex)
    {
        // Handle exceptions and return internal server error
        Console.WriteLine(ex);
        return new ObjectResult(new ArtworkResponse
        {
            Message = "An unexpected error occurred",
            Id = null
        })
        {
            StatusCode = StatusCodes.Status500InternalServerError
        };
    }
}
    }
}